imap/jmap_api.h

```diff
@@ -186,13 +186,31 @@ void jmap_getblob_ctx_fini(jmap_getblob_context_t *ctx);
 
 typedef int jmap_getblob_handler(jmap_req_t *req, jmap_getblob_context_t *ctx);
 
+// executed before request-specific handlers
+typedef void jmap_req_preproc(jmap_req_t *req);
+
 typedef struct {
     hash_table methods;
     json_t *server_capabilities;
     long limits[JMAP_NUM_LIMITS];
+    // internal state
     ptrarray_t getblob_handlers; // array of jmap_getblob_handler
+    ptrarray_t handlers; // array of (malloced) jmap_handlers
 } jmap_settings_t;
 
+
+enum jmap_handler_event {
+    JMAP_HANDLE_SHUTDOWN      = (1 << 1), /* executed when httpd is shutdown. req is NULL */
```

Comment:
Not that it matters, but is there a reason why this doesn't start at (1 << 0) ?

Reply:
No, that was just an oversight. Fixed.